### PR TITLE
parsing: rebuild antlr4 latex parser for 4.10

### LINKED ISSUE #19751: Remove python2 compatiblity support from LaTeX parser
https://github.com/sympy/sympy/blob/1eaefeb714cc05d87ea2f938813e5e859caa2b32/sympy/parsing/latex/LaTeX.g4#L18

I've tried several attempts to drop the line and rebuild the parser, but had not been successful due to some errors.

### LINKED ISSUE #22625: Changing antlr4 code generation language to Python3 produces import errors
It seems that a similar issue has been raised before (#19751), but since no progress has been made on that issue yet, I'm raising this again with some more details.

Currently, it seems that the [LaTeX.g4](https://github.com/sympy/sympy/blob/master/sympy/parsing/latex/LaTeX.g4) file still has the code generation language set to Python2:

https://github.com/sympy/sympy/blob/26110e2f95483393fd928072320734ef9cfa45c0/sympy/parsing/latex/LaTeX.g4#L16

Changing the language to Python3 and regenerating the  [latexlexer.py](https://github.com/sympy/sympy/blob/master/sympy/parsing/latex/_antlr/latexlexer.py) and [latexparser.py](https://github.com/sympy/sympy/blob/master/sympy/parsing/latex/_antlr/latexparser.py) files produces some import-related errors in latexlexer.py. For some reason, when you change the language to Python3, the `from antlr4 import *` import disappears from  latexlexer.py, leading to a number of errors. Adding the import back manually fixes the errors and passes or xfails all tests in [test_latex.py](https://github.com/sympy/sympy/blob/master/sympy/parsing/tests/test_latex.py). 

I'm not particularly familiar with ANTLR or the sympy parsing module, so I'm not sure what the cause of this issue is. Does anyone have any idea what is going on here? Could this possibly be a bug with ANTLR?

### LINKED ISSUE #23355: `sympy.parsing.latex.parse_latex` raising an error with `antlr4-python3-runtime==4.10`
`sympy==1.10.1`
`antlr4-python3-runtime==4.10`
Python 3.8.8 on Ubuntu 20.04LTS

When trying to use `sympy.parsing.latex.parse_latex` on a string, I get the following error:
```
>>> from sympy.parsing.latex import parse_latex 
>>> poly = '4y^2+4a^4-8ya^2'
>>> print(parse_latex(poly))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sofia/anaconda3/lib/python3.8/site-packages/sympy/parsing/latex/__init__.py", line 35, in parse_latex
    return _latex.parse_latex(s)
  File "/home/sofia/anaconda3/lib/python3.8/site-packages/sympy/parsing/latex/_parse_latex_antlr.py", line 73, in parse_latex
    lex = LaTeXLexer(stream)
TypeError: 'NoneType' object is not callable
```

No issue found with `antlr4-python3-runtime==4.9.3`

### PR DESCRIPTION
**Pull request by xrmx**
#### References to other Issues or PRs

Fixes #23355
Fixes #19751
Fixes #22625
#### Brief description of what is fixed or changed

antlr4 4.10 changed the ATN serialization format (https://github.com/antlr/antlr4/releases/tag/4.10) so that the current latex parser built with 4.7.2 does not work anymore with the latest antlr4-python3-runtime. This PR regenerate a python3 version of the latex grammar using latest antlr 4.10.

#### Other comments

Wondering about adding an optional dependency on a specific antlr4-python3-runtime version (https://setuptools.pypa.io/en/latest/userguide/dependency_management.html?highlight=install_requires#optional-dependencies)

#### Release Notes

* parsing
  * Regenerate autolev and latex parsers with antlr 4.10.1
  * Bump required antlr4 python runtime to 4.10.*

### CODE REVIEW

File: .github/workflows/runtests.yml [hunk 1]
```diff
@@ -148,7 +148,7 @@ jobs:
       - run: python -m pip install --upgrade pip
       - run: pip install mpmath gmpy2 matplotlib numpy scipy aesara ipython \
                          symengine cython llvmlite wurlitzer \
-                         autowrap numexpr 'antlr4-python3-runtime==4.7.*' \ 
+                         autowrap numexpr 'antlr4-python3-runtime==4.10' \
```
Comment (asmeurer) [inline .github/workflows/runtests.yml:151]: Should this also have `.*`?

**Comment by asmeurer** [comment]
Looks like the tests are failing here.

**Comment by asmeurer** [comment]
If we merge this, we need to make sure the antlr runtime pin is updated in the conda-forge recipe the next time SymPy is released.

File: .github/workflows/runtests.yml (hunk 1 above)
Comment (xrmx) [reply .github/workflows/runtests.yml:151]: The release is just 4.10 not 4.10.0

**Comment by xrmx** [comment]
Will take a look at the latex tests, the mailmap one looks unrelated.

**Comment by oscarbenjamin** [comment]
> the authors mailmap one looks unrelated.

Yes, it is: https://github.com/sympy/sympy/pull/23363#issuecomment-1100040502

File: .github/workflows/runtests.yml (hunk 1 above)
Comment (asmeurer) [reply .github/workflows/runtests.yml:151]: But what happens when 4.10.1 is released? Shouldn't it work as well?

Comment (xrmx) [reply .github/workflows/runtests.yml:151]: 4.10.1 has been released, can keep the wildcard.

**Comment by xrmx** [comment]
Puzzled why antlr 4.10 runtime could not parse the version from the serialized parser from the 4.10.1 jar. Took a bit to recognize it was autolev and not latex. I guess antlr target in setup.py should be extended to build the autolev stuff too.

**Comment by asmeurer** [comment]
It doesn't surprise me if there are some inconsistencies there, but they two parsers should definitely both be using the same functionality and tooling where possible.

**Comment by xrmx** [comment]
PR for building the autolev parser here https://github.com/sympy/sympy/pull/23416

**Comment by xrmx** [comment]
A quick update: with the last push latex tests are fine, autolev is still broken because IIRC for some reason the visitor is None and so nothing work, it well may be something silly but with all these dynamic imports the root causes tend to be well hidden :sweat_smile:

**Comment by oscarbenjamin** [comment]
I don't know anything about the antlr parsing so can't review this myself. Is this something that should go into the 1.11 release? If so please add the 1.11 milestone.

**Comment by eagleoflqj** [comment]
I think it would be good to upgrade this dependency in 1.11.

**Comment by asmeurer** [comment]
Agree. We also need to make sure to update the pin in the conda-forge recipe when the release happens.

**Pull request merged**